### ISSUE TITLE
make the waiting arguments configurable in clients initialization

### ISSUE THREAD

**Issue by cloudxxx8** [edgexfoundry/device-sdk-go#82]
The current implementation of clients/init.go:

```
	for i := 0; i < 50; i++ {
		if common.UseRegistry {
			if checkServiceAvailableByConsul(common.CurrentConfig.Clients[serviceId].Name) == true {
				return nil
			}
		} else {
			if checkServiceAvailableByPing(serviceId) == nil {
				return nil
			}
		}
		time.Sleep(2 * time.Second)
		common.LogCli.Debug(fmt.Sprintf("Checked %d times for %s availibility", i+1, serviceId))
	}
```

The 50 and 2 should be configurable.

**Comment by difince** [comment]
could this issue be assigned to me?

**Comment by cloudxxx8** [comment]
@difince Sure, please go ahead.  Thanks.

**Comment by difince** [comment]
I consider to add in the configuration.tolm files (section [Service]) these new properties: 
DependentSrvRetryCount = 50
DependentSrvRetryInterval = 2
Do you have another suggestion about the properties names ?

**Comment by difince** [comment]
//DependentSrvRetryCount is the max number of times the DS will try to connect to all dependent services.
	//If exceeded for even one dependent service, the DS will exit.
	DependentSrvRetryCount int
	//DependentSrvRetryInterval is the time in sec the DS will wait between each retry.
	DependentSrvRetryInterval int

**Comment by difince** [comment]
What about another approach. 
Instead of having in [Service] section common properties for all services/clients the DS depends on, we could have per each dependent client (Data, Metadata) this configuration:
```
[Clients]
[client.XXX]
......
RetryCount = 50
RetryInterval = 2
```

This way we will have a freedom to configure different values per each dependent client.

If the approach of having common properties for all clients is chosen - instead of "DependentSrvRetryCount"  that I suggested earlier - better to call them:
-  ClientRetryCount/ClientRetryInterval
or
-  ClientFailLimit/ClientFailWaitTime - just to be consistent with the [Registry] section

**Comment by cloudxxx8** [comment]
Although this way provides more flexibility, it becomes more complex from my point of view.  We would  make the config file longer.  It is nice for some circumstance, but the current dependency services might not need the different parameters.  I am fine with both approaches.

@tonyespy @tobiasmo1 @iain-anderson @SteveOss @lenny-intel how do you think?

**Comment by iain-anderson** [comment]
I don't think having individual retry configuration gains us much. Would prefer:

[Clients]
  RetryCount = 10
  RetryInterval = 1
  [Clients.Data]
    Host = "localhost"
...

**Comment by lenny-intel** [comment]
I agree with @iain-anderson .

**Comment by iain-anderson** [comment]
Actually... is this not what the existing configuration entries Service/Timeout and Service/ConnectRetries are for?

**Comment by difince** [comment]
yep.. 
We could use already existing property- Service/ConnectRetries (that currently is never used) and add new property as RetryInterval - this way not mixing between request "timeout" and request retry interval ?

**Comment by iain-anderson** [comment]
Would prefer we stick with Timeout so that it matches what the C SDK does

**Comment by cloudxxx8** [comment]
fixed by #198
